src/plugins/crypto.c

```diff
@@ -1433,196 +1327,138 @@ gboolean bd_crypto_luks_add_key (const gchar *device, const gchar *pass, const g
- * @error: (out) (optional): place to store error (if any)
- *
- * Returns: whether the key was successfully removed or not
- *
- * Either @pass or @key_file has to be != %NULL.
- *
- * Tech category: %BD_CRYPTO_TECH_LUKS-%BD_CRYPTO_TECH_MODE_REMOVE_KEY
- */
-gboolean bd_crypto_luks_remove_key_blob (const gchar *device, const guint8 *pass_data, gsize data_len, GError **error) {
-    struct crypt_device *cd = NULL;
-    gint ret = 0;
-    guint64 progress_id = 0;
-    gchar *msg = NULL;
-    GError *l_error = NULL;
-    gboolean success = FALSE;
-
-    msg = g_strdup_printf ("Started removing key from the LUKS device '%s'", device);
-    progress_id = bd_utils_report_started (msg);
-    g_free (msg);
-
-    ret = crypt_init (&cd, device);
-    if (ret != 0) {
-        g_set_error (&l_error, BD_CRYPTO_ERROR, BD_CRYPTO_ERROR_DEVICE,
-                     "Failed to initialize device: %s", strerror_l(-ret, c_locale));
-        bd_utils_report_finished (progress_id, l_error->message);
-        g_propagate_error (error, l_error);
-        return FALSE;
-    }
-
-    ret = crypt_load (cd, CRYPT_LUKS, NULL);
-    if (ret != 0) {
-        g_set_error (&l_error, BD_CRYPTO_ERROR, BD_CRYPTO_ERROR_DEVICE,
-                     "Failed to load device's parameters: %s", strerror_l(-ret, c_locale));
-        crypt_free (cd);
-        bd_utils_report_finished (progress_id, l_error->message);
-        g_propagate_error (error, l_error);
-        return FALSE;
-    }
-
-    success = _crypto_luks_remove_key (cd, pass_data, data_len, &l_error);
-    if (!success) {
-        crypt_free (cd);
+        g_set_error (&l_error, BD_CRYPTO_ERROR, BD_CRYPTO_ERROR_ADD_KEY,
+                     "Failed to add key: %s", strerror_l (-ret, c_locale));
         bd_utils_report_finished (progress_id, l_error->message);
         g_propagate_error (error, l_error);
         return FALSE;
     }
 
-    crypt_free (cd);
     bd_utils_report_finished (progress_id, "Completed");
     return TRUE;
 }
 
 /**
  * bd_crypto_luks_remove_key:
  * @device: device to add new key to
- * @pass: (nullable): passphrase for the @device or %NULL
- * @key_file: (nullable): key file for the @device or %NULL
+ * @context: key slot context (passphrase/keyfile/token...) to remove from this LUKS @device
  * @error: (out) (optional): place to store error (if any)
  *
  * Returns: whether the key was successfully removed or not
  *
- * Either @pass or @key_file has to be != %NULL.
- *
  * Tech category: %BD_CRYPTO_TECH_LUKS-%BD_CRYPTO_TECH_MODE_REMOVE_KEY
  */
-gboolean bd_crypto_luks_remove_key (const gchar *device, const gchar *pass, const gchar *key_file, GError **error) {
+gboolean bd_crypto_luks_remove_key (const gchar *device, BDCryptoKeyslotContext *context, GError **error) {
     struct crypt_device *cd = NULL;
     gint ret = 0;
     guint64 progress_id = 0;
     gchar *msg = NULL;
-    gboolean success = FALSE;
     gchar *key_buf = NULL;
     gsize buf_len = 0;
+    GError *l_error = NULL;
```

Comment:
Why this change to use a local error variable (only) in this function?

Reply:
Moved to a separate commit.